polybuild/polybuild.py

```diff
@@ -0,0 +1,358 @@
+
+
+class PolyBuilder:
+    def __init__(self, is_cxx):
+        self.meta = CompilerMeta(is_cxx, self.poly_find_dir(os.path.realpath(__file__)) + "/")
+
+    def poly_check_cxx(self, compiler: str) -> bool:
+        """
+        Checks if compiling a c++ or c program
+        """
+        if compiler.find("++") != -1:
+            return True
+        return False
+
+    def poly_find_dir(self, compiler_path: str) -> str:
+        """
+        Discover compiler install directory
+        Checks to see if the path is local directory, if not gives the entire path
+        """
+        last_slash: int = compiler_path.rfind("/")
+        if last_slash == -1:
+            return "."
+        return compiler_path[0:last_slash]
+
+    def poly_is_linking(self, argv) -> bool:
+        nonlinking_options = ["-E", "-fsyntax-only", "-S", "-c"]
+        for option in argv:
+            if option in nonlinking_options:
+                return False
+        return True
+
+    def poly_add_inst_lists(self, directory: str) -> Optional[List[str]]:
+        """
+        Adds a directory of lists to the instrumentation
+        """
+        dir_path = self.meta.compiler_dir + "../abi_lists/" + directory + "/"
+        file_list = []
+        if not os.path.exists(dir_path):
+            print(f"Error! {dir_path} not found!")
+            return None
+        dir_ents = os.listdir(dir_path)
+        for file in dir_ents:
+            if file != "." and file != "..":
+                file_list.append(dir_path + file)
+        return file_list
+
+    def poly_compile(self, bitcode_path: str, output_path: str, libs: List[str]) -> bool:
+        """
+        This function builds the compile command to instrument the whole program bitcode
+        """
+        compile_command = []
+        source_dir = self.meta.compiler_dir + "../lib/libTaintSources.a"
+        rt_dir = self.meta.compiler_dir + "../lib/libdfsan_rt-x86_64.a"
+        if self.meta.is_cxx:
+            compile_command.append("clang++")
+        else:
+            compile_command.append("clang")
+        compile_command += ["-pie", "-fPIC"]
+        optimize = os.getenv("POLYCLANG_OPTIMIZE")
+        if optimize is not None:
+            compile_command.append("-O3")
+        # -lpthread -Wl,--whole-archive libdfsan_rt-x86_64.a -Wl,--no-whole-archive libTaintSources.a -ldl -lrt -lstdc++
+        compile_command += ["-g", "-o", output_path, bitcode_path]
+        compile_command.append("-lpthread")
+        compile_command += ["-Wl,--whole-archive", rt_dir, "-Wl,--no-whole-archive", source_dir]
+        compile_command += ["-ldl", "-lrt"]
+        compile_command.append("-lstdc++")
+        for lib in libs:
+            if ".a" not in lib and ".o" not in lib:
+                compile_command.append("-l" + lib)
+            else:
+                compile_command.append(lib)
+        command = shlex.split(" ".join(shlex.quote(arg) for arg in compile_command))
+        ret_code = subprocess.call(command)
```

Comment:
I don't think the call to `shlex.split` is necessary. Instead, I think you should be able to just do `subprocess.call(compile_command)`.

Reply:
I will give it a try, but originally when i was doing this doing the subprocess call had issues when passing a string with additional double quotes inside of it. for example -DSOME_VAR="whatever", it would remove them. So I iterated through and created the new array by doing a join/split, but maybe i can try again.

Reply:
Done!